packages/grafana-data/src/transformations/transformers/concat.ts

```diff
@@ -0,0 +1,88 @@
+import { map } from 'rxjs/operators';
+
+import { DataTransformerID } from './ids';
+import { DataTransformerInfo } from '../../types/transformations';
+import { DataFrame, Field } from '../../types/dataFrame';
+import { ArrayVector } from '../../vector';
+
+export enum ConcatenateFrameNameMode {
+  Drop = 'drop',
+  FieldName = 'field',
+  Label = 'label',
+}
+
+export interface ConcatenateTransformerOptions {
+  frameNameMode?: ConcatenateFrameNameMode;
+  frameNameLabel?: string;
+}
+
+export const concatenateTransformer: DataTransformerInfo<ConcatenateTransformerOptions> = {
+  id: DataTransformerID.concatenate,
+  name: 'Concatenate fields',
+  description:
+    'Combine all fields into a single frame.  Values will be appended with undefined values if not the same length.',
+  defaultOptions: {
+    frameNameMode: ConcatenateFrameNameMode.FieldName,
+    frameNameLabel: 'frame',
+  },
+  operator: options => source =>
+    source.pipe(
+      map(dataFrames => {
+        if (!Array.isArray(dataFrames) || dataFrames.length < 2) {
+          return dataFrames; // noop with single frame
+        }
+        return [concatenateFields(dataFrames, options)];
+      })
+    ),
+};
+
+/**
+ * @internal only exported for tests
+ */
+export function concatenateFields(data: DataFrame[], opts: ConcatenateTransformerOptions): DataFrame {
+  let sameLength = true;
+  let maxLength = data[0].length;
+  const frameNameLabel = opts.frameNameLabel ?? 'frame';
+  let fields: Field[] = [];
```

Comment:
nitpick: newline here to separate the for statement from the preceding variable inits